org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/NodeModelBuilder.java

```diff
@@ -32,54 +30,58 @@
  */
 public class NodeModelBuilder {
 
-	private static class ArrayInterner<T> implements Interner<T[]> {
+	private static class ArrayInterner {
 
-		private static class ArrayAsList<T> extends AbstractList<T> implements RandomAccess {
-			final T[] array;
-			private int hashCode = -1;
+		private static class InternKey {
+			final EObject myGrammarElement;
+			final Object childGrammarElement;
+			
+			int hashCode = -1;
 
-			ArrayAsList(/* @Nullable */ T[] array) {
-				this.array = array;
+			InternKey(EObject myGrammarElement, Object childGrammarElement) {
+				this.myGrammarElement = myGrammarElement;
+				this.childGrammarElement = childGrammarElement;
 			}
-
-			@Override public int size() {
-				return array.length;
-			}
-
-			@Override /* @Nullable */ public T get(int index) {
-				return array[index];
+			
+			EObject[] joinElements() {
+				if(childGrammarElement instanceof EObject) {
+					return new EObject[] {myGrammarElement, (EObject) childGrammarElement};
+				} else {
+					return ObjectArrays.concat(myGrammarElement, (EObject[]) childGrammarElement);
```

Comment:
This uses refection internally. I'm slightly in favor of System.arrayCopy

Reply:
I couldn't measure any performance difference, I've looked at 'Array.newInstance' before and it's intrinsified (also the reason why it's often overrepresented in yourkit profiling results)

Reply:
Ok, same argumentation as below: If it doesn't show up somewhere, it's good enough :)